core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnIndexOf.java

```diff
@@ -0,0 +1,106 @@
+import gov.nist.secauto.metaschema.core.metapath.item.atomic.IAnyAtomicItem;
+import gov.nist.secauto.metaschema.core.metapath.item.atomic.IBooleanItem;
+import gov.nist.secauto.metaschema.core.metapath.item.atomic.IIntegerItem;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * /**
+ * Implements <a href=
+ * "https://www.w3.org/TR/xpath-functions-31/#func-index-of">fn:index-of</a>
+ * functions. This implementation does not implement the three-arg variant with
+ * collation at this time.
+ */
+public final class FnIndexOf {
+  @NonNull
+  private static final String NAME = "index-of";
+  @NonNull
+  static final IFunction SIGNATURE_TWO_ARG = IFunction.builder()
+      .name(NAME)
+      .namespace(MetapathConstants.NS_METAPATH_FUNCTIONS)
+      .deterministic()
+      .contextDependent()
+      .focusIndependent()
+      .argument(IArgument.builder()
+          .name("seq").type(IAnyAtomicItem.class)
+          .zeroOrMore()
+          .build())
+      .argument(IArgument.builder()
+          .name("search")
+          .type(IAnyAtomicItem.class)
+          .one()
+          .build())
+      .returnType(IIntegerItem.class)
+      .returnOne()
+      .functionHandler(FnIndexOf::executeTwoArg)
+      .build();
+
+  @SuppressWarnings("unused")
+  @NonNull
+  private static ISequence<IIntegerItem> executeTwoArg(@NonNull IFunction function,
+      @NonNull List<ISequence<?>> arguments,
+      @NonNull DynamicContext dynamicContext,
+      IItem focus) {
+    ISequence<?> seq = arguments.get(0).asSequence();
+    IAnyAtomicItem search = FunctionUtils.asType(arguments.get(1).getFirstItem(true));
+    
+    if (seq.size() == 0) {
+    	return ISequence.empty();
+    }
+    
+    return fnIndexOf(seq, search);
+  }
+
+  /**
+   * Determine if the string provided in the first argument contains the string in
+   * the second argument as a substring.
+   * <p>
+   * Based on the XPath 3.1 <a href=
+   * "https://www.w3.org/TR/xpath-functions-31/#func-index-of">fn:index-of</a>
+   * function.
+   *
+   * @param seq
+   *          the sequence to examine for matches to index
+   * @param search
+   *          the item to match and index if an equivalent item is in seq 
+   * @return a list of index numbers for matches in the seq
+   */
+  public static ISequence<IIntegerItem> fnIndexOf(@NonNull ISequence<? extends IItem> seq, @NonNull IAnyAtomicItem search) {
+	int index = 0;
+	ListIterator<? extends IItem> iterator = seq.listIterator();
+	List<IIntegerItem> indices = new ArrayList<>();
+	while(iterator.hasNext()) {
+		IItem item = iterator.next();
+		if(item.getValue().equals(search.getValue())) {
+			// Offset for Metapath indices that start from 1
+			indices.add(IIntegerItem.valueOf(index + 1));
+		}
+		index++;
+	}
+	return ISequence.ofCollection(indices);
+  }
```

Comment:
Blocker: I am not happy with this implementation but it took some trial and error. I know the pattern is often to go from complex types in the `execute...()` functions and use primite types. I spent 30 minutes with trial and error with `ArrayList<>()` and `ISequence.ofCollection()` with other tricks to consider how to build an array in this function and pass it back to the caller in `execute()`, but I think I am missing an understanding of the following:

How do I pass an array or Java List of primitives (such as `int`, not `Integer` to avoid boxing/unboxing) and then pass that to the caller so I can `ISequence.of()` or `ISequence.blah()` correctly? I spent a non-trivial amount of time on this experimenting with different approaches, and I could not figure it out.

If there is a better way to do this, I am open to it, this turned out to the be the simplest I thought with iterators and not making "down-convert" copies with primitive types, so you are more than welcome to say "no, A.J. I recommend a thoroughly different approach here."

Reply:
This is good. an ISequence is a List, so you can change its type to that.